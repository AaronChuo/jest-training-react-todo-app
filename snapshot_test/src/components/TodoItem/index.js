import React, { Component, PropTypes } from 'react';
import classnames from 'classnames';
import TodoTextInput from 'components/TodoTextInput';
import style from 'todomvc-app-css/index.css';

export default class TodoItem extends Component {
  static propTypes = {
    todo: PropTypes.object.isRequired,
    editTodo: PropTypes.func.isRequired,
    deleteTodo: PropTypes.func.isRequired,
    completeTodo: PropTypes.func.isRequired,
  };

  state = {
    editing: false,
  };

  handleDoubleClick = () => {
    this.setState({ editing: true });
  };

  handleSave = (id, text) => {
    if (text.length === 0) {
      this.props.deleteTodo({ id });
    } else {
      this.props.editTodo({ id, text });
    }
    this.setState({ editing: false });
  };

  render () {
    const { todo, completeTodo, deleteTodo } = this.props;

    let element;
    if (this.state.editing) {
      element = (
        <TodoTextInput
          text={todo.text}
          editing={this.state.editing}
          onSave={text => this.handleSave(todo.id, text)}
        />
      );
    } else {
      element = (
        <div className={style['view']}>
          <input
            className={style['toggle']}
            type="checkbox"
            checked={todo.completed}
            onChange={() => completeTodo({ id: todo.id })}
          />
          <label onDoubleClick={this.handleDoubleClick}>
            {todo.text}
          </label>
          <button className={style['destroy']} onClick={() => deleteTodo({ id: todo.id })} />
        </div>
      );
    }

    return (
      <li
        className={classnames({
          [style['completed']]: todo.completed,
          [style['editing']]: this.state.editing,
        })}
      >
        {element}
      </li>
    );
  }
}
