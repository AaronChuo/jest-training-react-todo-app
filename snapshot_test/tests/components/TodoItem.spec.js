import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import TodoItem from 'components/TodoItem';
import TodoTextInput from 'components/TodoTextInput';
import { todoList } from 'mock-data';

describe('<TodoItem />', () => {
  let handleEditTodo;
  let handleDeleteTodo;
  let handleCompleteTodo;
  let wrapper;

  beforeEach(() => {
    handleEditTodo = jest.fn();
    handleDeleteTodo = jest.fn();
    handleCompleteTodo = jest.fn();
    wrapper = shallow(
      <TodoItem
        todo={todoList[0]}
        editTodo={handleEditTodo}
        deleteTodo={handleDeleteTodo}
        completeTodo={handleCompleteTodo}
      />
    );
  });

  it('should render initial component', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('simulates double click to editing mode', () => {
    wrapper.find('.view label').simulate('doubleClick');
    expect(wrapper.state()).toMatchSnapshot();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('simulates save todo in editing mode', () => {
    const { id, text } = todoList[0];
    wrapper.setState({ editing: true });
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    wrapper.find(TodoTextInput).simulate('save', text);
    expect(handleEditTodo).toHaveBeenCalledWith({ id, text });
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('simulates save todo without text in editing mode', () => {
    const { id, text } = todoList[0];
    wrapper.setState({ editing: true });
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    wrapper.find(TodoTextInput).simulate('save', '');
    expect(handleDeleteTodo).toHaveBeenCalledWith({ id });
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('simulates change checkbox status to complete todo', () => {
    wrapper.find('.view .toggle').simulate('change');
    expect(handleCompleteTodo).toHaveBeenCalledWith({ id: todoList[0].id });
  });

  it('simulates click button to delete todo', () => {
    wrapper.find('.view .destroy').simulate('click');
    expect(handleDeleteTodo).toHaveBeenCalledWith({ id: todoList[0].id });
  });
});
