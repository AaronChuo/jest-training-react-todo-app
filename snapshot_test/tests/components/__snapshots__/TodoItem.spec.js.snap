// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem /> should render initial component 1`] = `
<li
  className=""
>
  <div
    className="view"
  >
    <input
      checked={false}
      className="toggle"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      onDoubleClick={[Function]}
    >
      Use Redux
    </label>
    <button
      className="destroy"
      onClick={[Function]}
    />
  </div>
</li>
`;

exports[`<TodoItem /> simulates double click to editing mode 1`] = `
Object {
  "editing": true,
}
`;

exports[`<TodoItem /> simulates double click to editing mode 2`] = `
<li
  className="editing"
>
  <TodoTextInput
    editing={true}
    onSave={[Function]}
    text="Use Redux"
  />
</li>
`;

exports[`<TodoItem /> simulates save todo in editing mode 1`] = `
<li
  className="editing"
>
  <TodoTextInput
    editing={true}
    onSave={[Function]}
    text="Use Redux"
  />
</li>
`;

exports[`<TodoItem /> simulates save todo in editing mode 2`] = `
<li
  className=""
>
  <div
    className="view"
  >
    <input
      checked={false}
      className="toggle"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      onDoubleClick={[Function]}
    >
      Use Redux
    </label>
    <button
      className="destroy"
      onClick={[Function]}
    />
  </div>
</li>
`;

exports[`<TodoItem /> simulates save todo without text in editing mode 1`] = `
<li
  className="editing"
>
  <TodoTextInput
    editing={true}
    onSave={[Function]}
    text="Use Redux"
  />
</li>
`;

exports[`<TodoItem /> simulates save todo without text in editing mode 2`] = `
<li
  className=""
>
  <div
    className="view"
  >
    <input
      checked={false}
      className="toggle"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      onDoubleClick={[Function]}
    >
      Use Redux
    </label>
    <button
      className="destroy"
      onClick={[Function]}
    />
  </div>
</li>
`;
